File: solutions/typescript/clock/2/clock.ts
export class Clock {
    private hours: number;
    private minutes: number;

    constructor(hours: number, minutes = 0) {
        this.hours = hours;
        this.minutes = minutes;
        this.normalize();
    }

    private normalize() {
        let totalMinutes = this.hours * 60 + this.minutes;
        totalMinutes = ((totalMinutes % (24 * 60)) + 24 * 60) % (24 * 60);
        this.hours = Math.floor(totalMinutes / 60);
        this.minutes = totalMinutes % 60;
    }

    public toString(): string {
        const pad = (num: number) => num.toString().padStart(2, '0');
        return `${pad(this.hours)}:${pad(this.minutes)}`;
    }

    public plus(minutes: number): Clock {
        return new Clock(this.hours, this.minutes + minutes);
    }

    public minus(minutes: number): Clock {
        return new Clock(this.hours, this.minutes - minutes);
    }

    public equals(other: Clock): boolean {
        return this.hours === other.hours && this.minutes === other.minutes;
    }
}

File: solutions/go/clock/2/clock.go
package clock

import "fmt"

type Clock int

const minutesPerDay = 24 * 60

func normalize(m int) Clock {
	m %= minutesPerDay
	if m < 0 {
		m += minutesPerDay
	}
	return Clock(m)
}

func New(h, m int) Clock {
	return normalize(h*60 + m)
}

func (c Clock) Add(m int) Clock {
	return normalize(int(c) + m)
}

func (c Clock) Subtract(m int) Clock {
	return normalize(int(c) - m)
}

func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", int(c)/60, int(c)%60)
}
